src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -216,6 +367,225 @@ DropOrphanedShards(bool waitForLocks)
 }
 
 
+/*
+ * StartNewOperationNeedingCleanup is be called by an operation to register
+ * for cleanup.
+ */
+OperationId
+StartNewOperationNeedingCleanup(void)
+{
+	CurrentOperationId = GetNextOperationId();
+
+	LockOperationId(CurrentOperationId);
+
+	return CurrentOperationId;
+}
+
+
+/*
+ * CompleteNewOperationNeedingCleanup is be called by an operation to signal
+ * completion. This will trigger cleanup of appropriate resources.
+ */
+void
+CompleteNewOperationNeedingCleanup(bool isSuccess)
+{
+	/*
+	 * As part of operation completion:
+	 * 1. Drop all resources of CurrentOperationId that are marked with 'CLEANUP_ALWAYS' policy and
+	 *     the respective cleanup records in seperate transaction.
+	 *
+	 * 2. For all resources of CurrentOperationId that are marked with 'CLEANUP_ON_FAILURE':
+	 *    a) If isSuccess = true, drop cleanup records as operation is nearing completion.
+	 *       As the operation is nearing successful completion. This is done as part of the
+	 *       same transaction so will rollback in case of potential failure later.
+	 *
+	 *    b) If isSuccess = false, drop resource and cleanup records in a seperate transaction.
+	 */
+
+	/* We must have a valid OperationId. Any operation requring cleanup
+	 * will call StartNewOperationNeedingCleanup.
+	 */
+	Assert(CurrentOperationId != INVALID_OPERATION_ID);
+
+	List *currentOperationRecordList = ListCleanupRecordsForCurrentOperation();
+
+	int removedShardCountOnComplete = 0;
+	int failedShardCountOnComplete = 0;
+
+	CleanupRecord *record = NULL;
+	foreach_ptr(record, currentOperationRecordList)
+	{
+		/* We only supporting cleaning shards right now */
+		if (record->objectType != CLEANUP_OBJECT_SHARD_PLACEMENT)
+		{
+			ereport(WARNING, (errmsg("Invalid object type %d for cleanup record ",
+									 record->objectType)));
+			continue;
+		}
+
+		if (record->policy == CLEANUP_ALWAYS ||
+			(record->policy == CLEANUP_ON_FAILURE && !isSuccess))
+		{
+			char *qualifiedTableName = record->objectName;
+			WorkerNode *workerNode = LookupNodeForGroup(record->nodeGroupId);
+
+			if (TryDropShardOutsideTransaction(qualifiedTableName, workerNode->workerName,
+											   workerNode->workerPort))
+			{
+				/*
+				 * Delete cleanup records outside transaction as:
```

Comment:
maybe add: When we do clean up in case of failure, we abort the transaction and therefore cannot delete records in the current transaction. We delete these records outside of the current transaction via a localhost connection.